### ISSUE TITLE
Add attention and transformer modules (MHA, GQA, RoPE, encoder/decoder layers, GRU)

### ISSUE THREAD

**Issue by xupremix** [xupremix/incin#101]
> Filed against the `feature_request.yml` template. Category: **Neural Network Layer / Module (`incin::nn`)**.

## Problem

There is no attention module. `docs/book/src/whats_not_finished.md`:

> **No transformer/attention modules.** The raw `scaled_dot_product_attention` operation exists; there is no `MultiHeadAttention` or `TransformerEncoderLayer` composed module, and no `GRU` alongside `LSTM`/`RNN`. Building a transformer means hand-composing it from primitives.

The prelude's layer exports confirm it: `Linear`, `Conv1d`, `Conv2d`, `BatchNorm2d`, `LayerNorm`, `RMSNorm`, `Embedding`, `Dropout`, `AvgPool2d`, `MaxPool2d`, `Flatten`, `RNN`, `RNNCell`, `LSTM`, `LSTMCell`, `Sequential`, the activations, and the losses. Nothing else.

The repository's own transformer proof shows the cost. `crates/incin/tests/transformer_block.rs` is a four-token, single-head block whose forward pass is hand-assembled:

```rust
// crates/incin/tests/transformer_block.rs - existing, works, and is the problem
fn forward(&self, input: Input) -> Result<Self::Output> {
    let query = self.query.forward(input.clone())?;
    let key = self.key.forward(input.clone())?;
    let value = self.value.forward(input.clone())?;
    let scores = query
        .matmul(&key.transpose(0isize, 1isize)?)?
        .mul_scalar(1.0 / (8.0_f32).sqrt())?;
    let attention = scores.softmax(1)?;
    let attended = attention.matmul(&value)?;
    let attention_residual = input + &self.projection.forward(attended)?;
    // ... feed-forward, residual
}
```

Its own doc comment records what it leaves out: "It deliberately leaves masking, normalization, dropout, and multi-head packing to a later composition layer." That layer is this issue. Note also that the proof is typed `Tensor<Dyn, Cpu, f32, Grad>`, not statically, which is a direct consequence of #100.

For the milestone goal of building state-of-the-art models quickly, this is the largest single gap: every current architecture is a stack of attention blocks, and the library asks the user to rebuild that stack from `matmul`, `transpose`, and `softmax` each time.

## Proposed API sketch

```rust
use incin::prelude::*;

type Backend = DefaultBackend;

const D_MODEL: usize = 512;
const N_HEADS: usize = 8;
const N_KV_HEADS: usize = 2;   // grouped-query attention; == N_HEADS is standard MHA
const SEQ: usize = 1024;

// ── proposed: multi-head attention with GQA as a parameter, not a separate module ──
let attention: MultiHeadAttention<D_MODEL, N_HEADS, N_KV_HEADS, Backend> =
    MultiHeadAttention::build(AttentionConfig {
        causal: true,                       // masking is configuration, not caller-constructed
        dropout: 0.0,
        position: PositionEncoding::Rotary { theta: 10_000.0 },
    })?;

let x: Tensor<s![2, SEQ, D_MODEL], Backend, f32, Grad> = /* ... */;
let y = attention.forward(x)?;              // Tensor<s![2, SEQ, D_MODEL], Backend, f32, Grad>

// ── proposed: a full pre-norm decoder layer ──
let layer: TransformerDecoderLayer<D_MODEL, N_HEADS, N_KV_HEADS, 2048, Backend> =
    TransformerDecoderLayer::build(TransformerConfig {
        norm: NormPlacement::Pre,           // explicit; not hardcoded
        feed_forward: FeedForwardKind::SwiGLU,
        ..Default::default()
    })?;

// ── proposed: a GPT-style stack, using the existing Sequential container ──
#[module]
struct Gpt {
    embed: Embedding<s![50_257, D_MODEL], Backend>,
    blocks: Sequential<[TransformerDecoderLayer<D_MODEL, N_HEADS, N_KV_HEADS, 2048, Backend>; 12]>,
    norm: RMSNorm<s![D_MODEL], Backend>,
    head: Linear<s![D_MODEL, 50_257], Backend>,
}
```

RoPE's frequency table is a **`Buffer`, not a `Param`** - it is state that must round-trip through a checkpoint but must not receive gradients or be touched by the optimizer. `Buffer<S, B, K>` already exists in `crates/incin-core/src/nn/param.rs:851` and is exported from the prelude, so this needs no new state concept:

```rust
// proposed - inside MultiHeadAttention
struct MultiHeadAttention</* ... */> {
    qkv: Linear</* ... */>,
    out: Linear</* ... */>,
    rope_cos: Buffer<s![SEQ, HEAD_DIM], B>,   // not a Param: no grad, no optimizer, but saved
    rope_sin: Buffer<s![SEQ, HEAD_DIM], B>,
}
```

## Shape and type-system invariants

This is the case the shape system exists for, so the constraints should be compile-time facts rather than runtime field checks:

| Invariant | Where it should be proven |
|---|---|
| `D_MODEL % N_HEADS == 0` | Compile time. `HEAD_DIM = D_MODEL / N_HEADS` is then a computed const, not a config field. |
| `N_HEADS % N_KV_HEADS == 0` | Compile time. GQA requires each KV head to serve a whole number of query heads. |
| `[B, T, D] -> [B, H, T, HEAD_DIM]` head split | Type-level reshape, derived from the const parameters rather than passed in. |
| Mask `[T, T]` against scores `[B, H, T, T]` | **#100.** This is the dependency, and it is why that RFC is sequenced first. |
| RoPE table extent >= runtime sequence length | Compile time where `SEQ` is static; typed error under `Dyn`. |

A mismatched head configuration must be a compile error, in the same style as `ValidMesh` in `crates/incin-core/src/dist/mesh.rs`, which uses `#[diagnostic::on_unimplemented]` to turn an invalid topology into a readable message. The same treatment applies here.

## Background

- **GQA is a parameter, not a module.** Multi-head attention, grouped-query attention, and multi-query attention differ only in how many KV heads exist (`N_HEADS`, some divisor, and `1` respectively). Three modules would be three copies of the same code.
- **RoPE cannot be an `Embedding`.** Learned absolute positions are an embedding lookup and `Embedding` already covers them. Rotary embeddings apply a position-dependent rotation to Q and K inside attention, so they are part of the attention module or they are nothing.
- **Norm placement is a real fork.** Pre-norm and post-norm have different training dynamics, and a library that hardcodes one is unusable for reproducing papers that used the other.
- **SwiGLU is a composition, not an activation.** It needs two projections and a gate, so it cannot be expressed as an entry in the existing `activation` module.

## How this interoperates

- **`Module` and state traversal.** Built on the existing `#[module]` macro and `Module<Input>` trait, so a transformer saves and loads through `ModelExt::save` / `ModelExt::load` with no special handling, and `collect_state` / `load_state` see the parameters and buffers correctly.
- **Catalog operations only.** Every module composes from operations already in `OPERATION_CATALOG`, so it runs on any backend advertising them. No fused kernel is required for correctness. #104 adds a fused path later as an optimization, selected under the module rather than replacing it.
- **The CUDA gap.** Running this on CUDA needs `embedding`, `tril`/`triu`, and `masked_fill` (#88) and a batched GEMM that is not a loop of 2D launches (#85). The modules land first and become fast when those close.
- **The existing proof.** `crates/incin/tests/transformer_block.rs` becomes the oracle: the module must produce identical output to the hand-composed block for the same weights, which checks the module against something that already works rather than only against itself.
- **Diagnostics.** Every const-parameter mismatch flows through `incin-diagnostics` and the `incin-lsp` rewrite layer, which is where a `typenum` chain becomes a decimal shape.

## Acceptance criteria

- A GPT-style decoder-only model is expressible in a readable amount of code using the facade, and trains on CPU.
- The same model runs on CUDA once #85 and #88 close, with no module-level changes.
- A mismatched head configuration (`D_MODEL % N_HEADS != 0`, or `N_HEADS % N_KV_HEADS != 0`) fails at compile time, covered by a compile-fail fixture.
- A transformer round-trips through save and load with exact state, RoPE buffers included, and the buffers receive no gradients.
- The module and the hand-composed block in `transformer_block.rs` produce identical outputs for identical weights.
- `docs/book/src/whats_not_finished.md` no longer lists missing attention modules, and `docs/book/src/transformer.md` shows the model end to end.

## Alternatives

- **Leave composition to users.** Keeps the core smaller and avoids taking a position on architecture details that change. It also means the library's answer to "build a transformer" stays "reassemble it from primitives", which is the friction this milestone exists to remove.
- **Ship `MultiHeadAttention` and RoPE only**, leaving the full layer to users. A reasonable smaller scope, and it still requires #100.
- **Runtime config fields instead of const generics.** Simpler signatures, and it converts every shape invariant above into a runtime panic. That is the opposite of the library's premise.

## Additional context

Relevant code: `crates/incin-core/src/nn/` (layer implementations), `crates/incin-core/src/nn/module.rs` (`Module`, `Sequential`, `TrainMode`), `crates/incin-core/src/nn/param.rs` (`Param`, `Buffer`), `crates/incin/tests/transformer_block.rs`, `crates/incin-core/src/dist/mesh.rs` (`ValidMesh` as the in-tree pattern for a compile-time divisibility proof with a readable diagnostic), `docs/book/src/transformer.md`.

Depends on #100 for masking through the typed path. Depends on #88 and #85 for CUDA execution. #104 adds the fused kernel under this surface.
